bfabric/bfabric_config.py

```diff
@@ -27,73 +30,146 @@ class BfabricConfig:
-        if key not in [
-            "_PASSWD",
-            "_LOGIN",
-            "_WEBBASE",
-            "_APPLICATION",
-            "_JOB_NOTIFICATION_EMAILS",
-        ]:
-            continue
-
-        # In case of multiple definitions, the first rule counts!
-        if key not in values:
-            if key in ["_APPLICATION"]:
-                try:
-                    values[key] = json.loads(value)
-                except json.JSONDecodeError as e:
-                    raise ValueError(
-                        f"While reading {file_path}. '{key}' is not a valid JSON string."
-                    ) from e
-            else:
-                # to make it downward compatible; so we replace quotes in login and password
-                values[key] = value.replace('"', "").replace("'", "")
+    logger.info(f"Reading configuration from: {config_path}")
+
+    # Read the config file
+    config_dict = yaml.safe_load(Path(config_path).read_text())
+
+    # config = ConfigParser()
+    # config.read(config_path)
+    # config_dict = {s: dict(config.items(s)) for s in config.sections()}
+
+    if "GENERAL" not in config_dict:
+        raise IOError("Config file must have a general section")
+    if 'default_config' not in config_dict['GENERAL']:
+        raise IOError("Config file must provide a default environment")
+    config_env_default = config_dict['GENERAL']['default_config']
+
+    # Determine which environment we will use
+    # By default, use the one provided by config_env
+    if config_env is None:
+        # Try to find a relevant
+        config_env = os.getenv("BFABRICPY_CONFIG_ENV")
+        if config_env is None:
+            logger.log(20, "BFABRICPY_CONFIG_ENV not found, using default environment " + config_env_default)
+            config_env = config_env_default
         else:
-            logger.warning(f"While reading {file_path}. '{key}' is already set.")
-
-    args = dict(
-        base_url=values.get("_WEBBASE"),
-        application_ids=values.get("_APPLICATION"),
-        job_notification_emails=values.get("_JOB_NOTIFICATION_EMAILS"),
-    )
-    config = BfabricConfig(**{k: v for k, v in args.items() if v is not None})
-    if "_LOGIN" in values and "_PASSWD" in values:
-        auth = BfabricAuth(login=values["_LOGIN"], password=values["_PASSWD"])
+            logger.log(20, "found BFABRICPY_CONFIG_ENV = " + config_env)
     else:
+        logger.log(20, "config environment specified explicitly as " + config_env)
+
+    if config_env not in config_dict:
+        raise IOError("The requested config environment", config_env, "is not present in the config file")
+
+    return config_env, config_dict[config_env]
+
+def _have_all_keys(d: dict, l: list) -> bool:
+    """True if all elements in list l are present as keys in dict d, otherwise false"""
+    return all([k in d for k in l])
+
+def _parse_dict(d: dict, mandatory_keys: list, optional_keys: list = None, error_prefix: str = None):
```

Comment:
```suggestion
def _parse_dict(d: dict, mandatory_keys: list, optional_keys: list = None, error_prefix: str = " "):
```
the code would have failed before if it were `None`

Reply:
Indeed, my bad